File: lib/core/models/student.dart
import 'package:json_annotation/json_annotation.dart';

part 'student.g.dart';

/// 性别
enum Gender {
  @JsonValue('male') male,
  @JsonValue('female') female,
  @JsonValue('other') other;

  String get label {
    switch (this) {
      case Gender.male:
        return '男';
      case Gender.female:
        return '女';
      case Gender.other:
        return '其他';
    }
  }
}

@JsonSerializable(fieldRename: FieldRename.snake, explicitToJson: true)
class Student {
  final int id;
  final String name;
  final Gender gender;
  final int age;
  final double heightCm;
  final double targetWeightKg;
  final DateTime startDate;
  final String notes;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Student({
    required this.id,
    required this.name,
    required this.gender,
    required this.age,
    required this.heightCm,
    required this.targetWeightKg,
    required this.startDate,
    required this.notes,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Student.fromJson(Map<String, dynamic> json) => _$StudentFromJson(json);
  Map<String, dynamic> toJson() => _$StudentToJson(this);
}

/// 新建/更新学员请求体（不含 id 与时间戳）
@JsonSerializable(fieldRename: FieldRename.snake, explicitToJson: true)
class StudentInput {
  final String name;
  final Gender gender;
  final int age;
  final double heightCm;
  final double targetWeightKg;
  final DateTime startDate;
  final String notes;

  const StudentInput({
    required this.name,
    required this.gender,
    required this.age,
    required this.heightCm,
    required this.targetWeightKg,
    required this.startDate,
    this.notes = '',
  });

  factory StudentInput.fromJson(Map<String, dynamic> json) =>
      _$StudentInputFromJson(json);
  Map<String, dynamic> toJson() => _$StudentInputToJson(this);
}

File: backend/app/schemas/student.py
"""学员档案 Schema。"""
from datetime import date, datetime
from pydantic import BaseModel, Field


class StudentBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=50)
    gender: str = Field(..., pattern="^(male|female|other)$")
    age: int = Field(..., ge=1, le=150)
    height_cm: float = Field(..., gt=0)
    target_weight_kg: float = Field(..., gt=0)
    start_date: date
    notes: str = ""


class StudentCreate(StudentBase):
    """新建学员请求体。"""
    pass


class StudentUpdate(BaseModel):
    """更新学员（全部可选，PATCH 语义）。"""
    name: str | None = Field(None, min_length=1, max_length=50)
    gender: str | None = Field(None, pattern="^(male|female|other)$")
    age: int | None = Field(None, ge=1, le=150)
    height_cm: float | None = Field(None, gt=0)
    target_weight_kg: float | None = Field(None, gt=0)
    start_date: date | None = None
    notes: str | None = None


class StudentOut(StudentBase):
    """学员响应体。"""
    id: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
